### DISCUSSION TITLE
submit calculated data and push to new component

### DISCUSSION THREAD

**Discussion by d-opri** [neuefische/questions#496] [Web] [answered]
moin moin, 

my form is supposed to take the users input and caluclate the a balance for all added participants with the total amount, which works. but now i am trying to have the results displayed on a new page with a different component, and that component is then meant to submit the data and create a new bill with a details page. i can display the results when i keep all components in one, but the moment i try to break it up it goes wrong. this is just the form, i also have an add-bill page where i import the function and fetch the data then use router push to navigate to the result details page where i import the component to display the data with use swr. its a lot of code so i thought it best to go through it in chunks to see where i am going wrong. 

thank u for any help ( ͒  ́ඉ .̫ ඉ ̀ ͒)

```
export default function BillForm({ onSubmit }) {
  const [title, setTitle] = useState("");
  const [amount, setAmount] = useState("");
  const [participants, setParticipants] = useState([
    {
      name: "",
      id: uuid(),
      paid: 0,
    },
  ]);
  const [balance, setBalance] = useState({
    name: "",
    balance: 0,
  });

  function handleAddParticipant(event) {
    / ... /
  }

  function handleDeleteParticipant(id) {
    / ... /
  }

  function handleParticipantChange(event, index) {
    / ... /
  }

  async function calculateSplit(amount, participants) {
    const evenSplit = amount / participants.length;
    const newBalance = participants.map((participant) => {
      const share = evenSplit - participant.paid;
      return {
        name: participant.name,
        balance: balance > 0 ? ` owes ${share} $` : ` is owed ${-share} $`,
      };
    });
    return setBalance(newBalance);
  }

  function handleSubmit(event) {
    event.preventDefault();

    const formData = {
      title,
      amount,
      participants,
      balance,
    };
    const data = Object.fromEntries(Object.entries(formData));

    onSubmit(data);
  }

  return (
    <>
      <StyledLink href='/'>X</StyledLink>
      <Title>Create New Bill</Title>

      <Container onSubmit={handleSubmit}>
        <label aria-label='input to add participants' htmlFor='title'>
          Title
          <input
            type='text'
            onChange={(event) => setTitle(event.target.value)}
            value={title}
            name='title'
            id='title'
          />
        </label>

        <label aria-label='amount' htmlFor='amount'>
          Amount
          <input
            type='number'
            onChange={(event) => setAmount(event.target.value)}
            value={amount}
            name='amount'
            placeholder='00.00'
            id='amount'
          />
        </label>
        <button type='button' onClick={handleAddParticipant}>
          Add a participant
        </button>

        {participants.map((participant, index) => (
          <li key={index}>
            <label htmlFor={index}>
              <input
                type='text'
                placeholder='Add Name'
                id={index}
                name='name'
                value={participant.name}
                onChange={(event) => handleParticipantChange(event, index)}
              />
            </label>
            <label htmlFor={index}>
              Paid:
              <input
                type='number'
                id={index}
                name='paid'
                value={participant.paid}
                onChange={(event) => handleParticipantChange(event, index)}
              />
            </label>
            <button
              type='button'
              onClick={() => handleDeleteParticipant(participant.id)}
            >
              X
            </button>
          </li>
        ))}

        <button type='submit' onClick={calculateSplit}>
          Calculate Split
        </button>
      </Container>
    </>
  );
}
```

**Comment by mbosselmann** [comment]
Hi Danie,
can you add a link to your current branch please? :) 
Thank you ❤️

**Reply by d-opri** [reply]
yes of course huh seems the link didnt work? (https://github.com/d-opri/Kroko-Capstone/pull/7)

**Comment by mbosselmann** [comment, marked as answer]
Hi Danie,
this took me longer than expected, but now I have a working solution for you. As recap: The basic idea is that you want the data to be entered in the form to be displayed in your `BillDetails` component. 
I have added the `BillDetails` component on your index page under your Activitiy component as a test.

Currently you are working with multiple states in your form and you can access them directly to display the data on the same page with your `BillDetails`component. To have this data also available on another page, you need to store this data one level higher in the `_app.js` in a state:

**_app.js**
```js
import { useState } from "react";
import GlobalStyle from "@/styles";
import Head from "next/head";

export default function App({ Component, pageProps }) {
  const [bill, setBill] = useState(null);

  function addBill(newBill) {
    setBill(newBill);
  }

  return (
    <>
      <GlobalStyle />
      <Head>
        <title>Capstone Project</title>
      </Head>
      <Component {...pageProps} bill={bill} addBill={addBill} />
    </>
  );
}

```
The keyword for this is **lifting state up**. [See React beta docs fur further information.](https://beta.reactjs.org/learn/sharing-state-between-components).

To your pages you can pass this state and also other functions, as I did.

Your `BillForm` component needs the update function to be able to give the data one level up so you can pass the data to the other page, in my example I give the bill data to the `HomePage`:

**HomePage or index.js:**
```js
import BillDetails from "@/components/BillDetails.js";
import Activity from "../components/activity";

export default function HomePage({ bill }) {
  return (
    <>
      <Activity />
      {bill && (
        <BillDetails title={bill.title} total={bill.amount}>
          {bill.participants.map((participant) => {
            <div key={participant.id}>
              {participant.name}
              {participant.balance}
            </div>;
          })}
        </BillDetails>
      )}
    </>
  );
}
```

(You should move your `Activitiy` component to the components folder as this is not a page.)

**create-bill/index.js**

```js

export default function BillForm({ addBill }) {
  ...

  function handleSubmit(event) {
    event.preventDefault();
    const fairShare = amount / participants.length;
    const newBalance = participants.map((participant) => {
      const balance = fairShare - participant.paid;
      return {
        name: participant.name,
        balance: balance > 0 ? ` owes ${balance} $` : ` is owed ${-balance} $`,
      };
    });
    setResults(newBalance);
    const newBill = { title, amount, participants: newBalance };
    addBill(newBill);
  }

...
}

```

In your form component I created a new bill object that gets all the form data and use the `addBill` function to store the data in my bill state in my `_app.js` file.
As for now you only want to display one bill I created an `newBill` object. If you want to display more than one bill, you need to adjust the `addBill` function in the `_app.js`.  Maye our recap project 5 (art gallery app) is helpful for you for this case. :)

I hope that this already helps you. :)

**Reply by d-opri** [reply]
thanks a lot for all the mühe, i wouldnt have gotten anywhere without it! (人*´∪`)♪ тнайк　чоц♪(´∪`*人)
